runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java

```diff
@@ -0,0 +1,231 @@
+      // to have a checkpoint to potentially resume from.
+      residual = tracker.checkpoint();
+    } else {
+      residual = forcedCheckpoint;
+    }
+    return new Result(residual, cont);
+  }
+
+  private class ProcessContext extends DoFn<InputT, OutputT>.ProcessContext {
+    private final WindowedValue<InputT> element;
+    private final TrackerT tracker;
+
+    private int numOutputs;
+    private RestrictionT checkpoint;
+
+    private Future<?> scheduledCheckpoint;
+
+    public ProcessContext(WindowedValue<InputT> element, TrackerT tracker) {
+      fn.super();
+      this.element = element;
+      this.tracker = tracker;
+
+      this.scheduledCheckpoint =
+          executor.schedule(
+              new Runnable() {
+                @Override
+                public void run() {
+                  initiateCheckpoint();
+                }
+              },
+              maxDuration.getMillis(),
+              TimeUnit.MILLISECONDS);
+    }
+
+    @Nullable
+    RestrictionT extractCheckpoint() {
+      scheduledCheckpoint.cancel(true);
+      try {
+        scheduledCheckpoint.get();
+      } catch (InterruptedException | ExecutionException | CancellationException e) {
+        // Ignore
+      }
+      synchronized (this) {
+        return checkpoint;
+      }
+    }
+
+    @Override
+    public InputT element() {
+      return element.getValue();
+    }
+
+    @Override
+    public <T> T sideInput(PCollectionView<T> view) {
+      return sideInputReader.get(
+          view,
+          view.getWindowingStrategyInternal()
+              .getWindowFn()
+              .getSideInputWindow(Iterables.getOnlyElement(element.getWindows())));
+    }
+
+    @Override
+    public Instant timestamp() {
+      return element.getTimestamp();
+    }
+
+    @Override
+    public PaneInfo pane() {
+      return element.getPane();
+    }
+
+    @Override
+    public PipelineOptions getPipelineOptions() {
+      return pipelineOptions;
+    }
+
+    @Override
+    public void output(OutputT output) {
+      outputWithTimestamp(output, element.getTimestamp());
+    }
+
+    @Override
+    public void outputWithTimestamp(OutputT value, Instant timestamp) {
+      output.outputWindowedValue(value, timestamp, element.getWindows(), element.getPane());
+      noteOutput();
+    }
+
+    @Override
+    public <T> void sideOutput(TupleTag<T> tag, T value) {
+      sideOutputWithTimestamp(tag, value, element.getTimestamp());
+    }
+
+    @Override
+    public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T value, Instant timestamp) {
+      output.sideOutputWindowedValue(
+          tag, value, timestamp, element.getWindows(), element.getPane());
+      noteOutput();
+    }
+
+    private void noteOutput() {
+      ++numOutputs;
+      if (numOutputs >= maxNumOutputs) {
+        initiateCheckpoint();
+      }
+    }
+
+    private synchronized void initiateCheckpoint() {
```

Comment:
I couldn't come up with cleaner code for this, but I moved the accesses to checkpoint closer together, and added a bunch of comments.